crates/dojo-core/src/database.cairo

```diff
@@ -70,8 +82,50 @@ fn all(
                 entities.append(value);
             },
             Option::None(_) => {
-                break (all_ids.span(), entities.span());
+                break entities.span();
             }
         };
     }
 }
+
+/// Returns entries on the given keys.
+/// # Arguments
+/// * `class_hash` - The class hash of the contract.
+/// * `component` - The component to get the entries from.
+/// * `partition` - The partition of the component to get the entries from.
+/// * `key` - The key of the entries to get.
+/// * `length` - The length of the entries.
+fn get_by_key(
+    class_hash: starknet::ClassHash, component: felt252, partition: felt252, key: felt252, length: usize, layout: Span<u8>
+) -> (Span<felt252>, Span<Span<felt252>>) {
+        let table = {
+        if partition == 0.into() {
+            component
+        } else {
+            let mut serialized = ArrayTrait::new();
+            component.serialize(ref serialized);
+            partition.serialize(ref serialized);
+            let hash = poseidon_hash_span(serialized.span());
+            hash.into()
+        }
+    };
+
+    let all_ids = index::get_by_key(0, table, key);
+    (all_ids.span(), get_by_ids(class_hash, table, all_ids.span(), length, layout))
+}
+
+/// Set, but with writing keys to the appropriate indexes
+/// # Arguments
+/// * `class_hash` - The class hash of the contract.
+/// * `table` - The table to set the entries to.
+/// * `id` - The id of the entry to set.
+/// * `offset` - The offset of the entry to set.
+/// * `value` - The value of the entry to set.
+/// * `keys` - The keys of the entry to set in the index.
+fn set_with_keys(
+    class_hash: starknet::ClassHash, table: felt252, id: felt252, offset: u8, value: Span<felt252>, layout: Span<u8>, keys: Span<felt252>, keys_layout: Span<u8>
+
+) {
+    set(class_hash, table, id, offset, value, layout);
+    index::create_with_keys(0, table, id, keys, keys_layout);
```

Comment:
Does calling this on `set` create redundant keys?

Reply:
It shouldn't, and doesn't in tests as in set indexing is disabled by default

Reply:
This is how I understand this feature we have here:

## 1. Dojo Key Generation

When an object is stored by dojo, a unique key is generated. We refer to these keys as `dojo keys`.

## 2. Tagging System

Whenever a new tag is introduced, a specific `dojo key` is generated for it. This isn't a direct link to the object, but to an array:

- This array houses `dojo keys` of all objects that bear that tag.

## 3. Adding New Tags

On introducing a new object with a fresh tag (one that hasn’t been used previously):

   a. A `dojo key` is generated for the object.
   
   b. Another `dojo key` is dedicated to this fresh tag. This key leads to an array, and initially, the array will contain just the `dojo key` of the new object.

## 4. Using Existing Tags

For an object with an already existing tag:

   a. A single new `dojo key` is produced for the object.
   
   b. The pre-existing array (linked with a `dojo key` for that tag) gets updated. The `dojo key` of the new object is appended, marking this object with that tag.

Reply:
Exactly, with a implementation detail that this array isn't an ArrayTrait but rather just a len, and hashes for each index below that are addresses in storage of `dojo keys`.